src/optim/FishLeg/fishleg.py

```diff
@@ -226,53 +293,85 @@ def update_aux(self) -> None:
         where :math:`\\theta` is the parameters of model, :math:`\lambda` is the
         auxliarary parameters.
         """
-        data_x, _ = next(iter(self.aux_dataloader))
-        data_x.to(self.device)
+        
+        data = next(iter(self.aux_dataloader))
+        data = self._prepare_input(data) 
 
         self.aux_opt.zero_grad()
         with torch.no_grad():
-            pred = self.draw(self.model, data_x)
-            g2 = 0.0
-            for group in self.param_groups:
-                for p in group["params"]:
-                    g2 += p.grad.data.norm(p=2) ** 2
-            g2 = torch.sqrt(g2)
-
-        aux_loss = 0.0
+            self.store_g = False
+            samples = self.draw(self.model, data)
+            self.store_g = True
+        
+
+        g2 = 0.0
         for group in self.param_groups:
-            name = group["name"]
+            name = group['name']
+            for i, (p, para_name) in enumerate(zip(group["params"], group['order'])):
+                grad = p.grad.data
+                g2 = g2 + torch.sum(grad * grad)
+                group['grad'][i].copy_(grad) 
+        
+        g_norm = torch.sqrt(g2) 
+
+        self.zero_grad()
+        # How to better implement this?
+        # The hook is not updated here, locally, only the gradient to the parameters g.grad is updated
+        self.store_g = False
+        self.nll(self.model, samples).backward()
+        self.store_g = True
```

Comment:
What is this doing here? Does this store_g parameter affect the model somehow?

Reply:
The store_g need to be True for the forward storage of input (activated inside `.nll`) and backward storage of gradient (activated by `.backward()`) to be executed. I don't want to execute them, because `Qg()` will still use gradient outside the aux_update step. This backward step will only update the gradient attach to parameter.grad, but not layer._a and layer._g.